pandas/core/tools/datetimes.py

```diff
@@ -60,7 +66,35 @@ def _maybe_cache(arg, format, cache, convert_listlike):
     return cache_array
 
 
-def _convert_and_box_cache(arg, cache_array, box, errors, name=None):
+def _box_as_indexlike(dt_array: ArrayLike, tz=None, name=None) -> IndexLike:
+    """
+    Properly boxes the ndarray of datetimes to DatetimeIndex
+    if it is possible or to generic Index instead
+
+    Parameters
+    ----------
+    dt_array: 1-d array
+        array of datetimes to be boxed
+    tz : object
+        None or 'utc'
+    name : string, default None
+        Name for a resulting index
+
+    Returns
+    -------
+    result : datetime of converted dates
+        - DatetimeIndex if convertible to sole datetime64 type
+        - general Index otherwise
+    """
+    from pandas import DatetimeIndex, Index
+    if is_datetime64_dtype(dt_array):
+        return DatetimeIndex(dt_array, tz=tz, name=name)
+    return Index(dt_array, name=name)
+
+
+def _convert_and_box_cache(arg: IntFltStrDateLisTuplArrSer,
+                           cache_array: SeriesLike,
+                           box: bool, name=None) -> IndexLikeOrNdarray:
```

Comment:
just do the Union[Index, np.ndarray]  here

Reply:
I use `Union[ABCIndex, np.ndarray]` here to not generate cyclic imports that will occur during `import Index`(out of function)